service/common/src/main/java/org/apache/polaris/service/catalog/iceberg/CatalogHandlerUtils.java

```diff
@@ -433,39 +450,203 @@ protected TableMetadata commit(TableOperations ops, UpdateTableRequest request)
+      }
+    }
+
+    List<MetadataUpdate> updateToRemoveSnapshot = new ArrayList<>();
+    Long snapshotId = base.ref(updateRefName).snapshotId(); // current tip of the given branch
+    // ensure this branch has the latest sequence number.
+    long expectedSequenceNumber = base.lastSequenceNumber();
+    Set<Long> snapshotsToRemove = new LinkedHashSet<>();
+    while (snapshotId != null && !Objects.equals(snapshotId, expectedCurrentSnapshotId)) {
+      Snapshot snap = base.snapshot(snapshotId);
+      if (expectedSequenceNumber != snap.sequenceNumber()) {
+        break;
+      }
+      if (!isRollbackSnapshot(snap) || idsToRetain.contains(snapshotId)) {
+        // Either encountered a non no-op snapshot or the snapshot is being referenced by any other
+        // reference either by branch or a tag.
+        break;
+      }
+      snapshotsToRemove.add(snap.snapshotId());
+      snapshotId = snap.parentId();
+      // we need continuous sequence number to correctly rollback
+      expectedSequenceNumber--;
+    }
+
+    boolean wasExpectedSnapshotReached = Objects.equals(snapshotId, expectedCurrentSnapshotId);
+    updateToRemoveSnapshot.add(new MetadataUpdate.RemoveSnapshots(snapshotsToRemove));
+    return wasExpectedSnapshotReached ? updateToRemoveSnapshot : null;
+  }
+
+  private boolean isRollbackSnapshot(Snapshot snapshot) {
+    // Only Snapshots with {@ROLLBACKABLE_REPLACE_SNAPSHOT} are allowed to be rollback.
+    return DataOperations.REPLACE.equals(snapshot.operation())
+        && PropertyUtil.propertyAsString(snapshot.summary(), CONFLICT_RESOLUTION_ACTION, "")
+            .equalsIgnoreCase("rollback");
+  }
+
+  private MetadataUpdate.SetSnapshotRef findSetSnapshotRefUpdate(UpdateTableRequest request) {
+    int total = 0;
+    MetadataUpdate.SetSnapshotRef setSnapshotRefUpdate = null;
+    // find the SetRefName snapshot update
+    for (MetadataUpdate update : request.updates()) {
+      if (update instanceof MetadataUpdate.SetSnapshotRef) {
+        total++;
+        setSnapshotRefUpdate = (MetadataUpdate.SetSnapshotRef) update;
+      }
+    }
+
+    // if > 1 assertion for refs, then it's not safe to rollback, make this Noop.
+    return total != 1 ? null : setSnapshotRefUpdate;
+  }
+
+  private MetadataUpdate.AddSnapshot findAddSnapshotUpdate(UpdateTableRequest request) {
+    int total = 0;
+    MetadataUpdate.AddSnapshot addSnapshot = null;
+    // find the SetRefName snapshot update
+    for (MetadataUpdate update : request.updates()) {
+      if (update instanceof MetadataUpdate.AddSnapshot) {
+        total++;
+        addSnapshot = (MetadataUpdate.AddSnapshot) update;
+      }
+    }
+
+    // if > 1 assertion for addSnapshot, then it's not safe to rollback, make this Noop.
+    return total != 1 ? null : addSnapshot;
+  }
+
+  private TableMetadata setAppropriateLastSeqNumber(TableMetadata newBase) {
+    // TODO: Get rid of the reflection call once TableMetadata have API for it.
```

Comment:
If we merge this, let's file an issue to track this TODO

Reply:
done : https://github.com/apache/polaris/issues/1917 assigned to myself